### PR TITLE
JavaScript: New ESLint Fathead

### PR DESCRIPTION
**Pull request by zodiac403**
## Description of new Instant Answer, or changes

Creating my first fathead. IA page can be found here: https://duck.co/ia/view/eslint

The **output.txt** does not cover all articles yet. But it would be great to get some feedback, if my approach is correct or if I missed anything.

I followed [these](steps) to create a pull request. Two questions remained open:
- Is the *master* branch the correct branch to merge to?
- How does the IA move from *Planning* to *Development*?

**Any feedback is welcome. If there is a better platform to discuss these questions, please let me know.**

## Related Issues and Discussions

none

## People to notify

none

## Testing & Review
To be completed by Language Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [x] Title follows correct format (Specifies Instant Answer + Purpose)
- [x] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [x] Instant Answer page is correctly filled out and contains:
    - One topic for the Search Space Language (Java, Python, Scala, Ruby, etc.)
    - One topic from: Reference, Help, Libraries, Tools
        - Documentation Fatheads are considered "Reference"
    - Description, Data source, and 2+ example queries
    - Perl Module (e.g. "DDG::Fathead::PerlDoc" -- we only need a name, not an actual file)
    - Source Name (for "More at <source_name>" link)
    - Source Domain (must contain http:// or https:// -- can be the same as Data Source)
    - Source Info (used as Subtitle for each Article -- usually matches the IA Name)
    - 'Skip Abstract' is checked off
    - Source ID (ping @moollaza to assign one, once Fathead is ready for Beta deploy)

**Code**
- [x] Uniformly indented, well commented
- [x] Fetch.sh and Parse.sh run without errors
- [ ] Output contains no blank lines, or multi-line entries
- [ ] Fathead Tests are passing (run `$ duckpan test <fathead_id>`)
    - Tester should report any failures

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/eslint

### CODE REVIEW

**Comment by daxtheduck** [comment]
@zodiac403 Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by moollaza** [comment]
> Is the master branch the correct branch to merge to?

Yup 👍 

> How does the IA move from Planning to Development?

I've moved it for you. If you open a PR and link your IA page, it should happen automatically

File: lib/fathead/javascript_eslint/doc/notes.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Notes
```
Comment (moollaza) [inline lib/fathead/javascript_eslint/doc/notes.md]: I would just move the important parts of this into the README, and make that a markdown file.

Any reason you've created /doc/notes.md?

Comment (zodiac403) [inline lib/fathead/javascript_eslint/doc/notes.md]: @moollaza thx for your quick reply. I took a look at the other fatheads in this repo. None of them had a MD doc at top-level, so I moved to doc/.

No problem, I'll move that.

**Comment by bfmags** [comment]
@zodiac403 looking good! let me know when it's read to review

**Comment by zodiac403** [comment]
@bfmags at the moment I have some struggles with the layout. I described them here: https://forum.duckduckhack.com/t/fathead-display-how-to-set-more-at-and-subtitle/1575
Probably you have an idea?

**Comment by zodiac403** [comment]
@bfmags OK, I think, I got a minimum viable version together.

The abstract could be more detailed, but this will require some major changes to fetch and parse. It would be great to see this one up and running, before I approach this topic.

Let me know what you think about it.

**Comment by moollaza** [comment]
@zodiac403 please change the dir in `lib/fathead/` to `eslint`. The ID and dir name must be identical 👍

File: lib/fathead/eslint/lib/convert.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * and convert to a pretty rendered HTML snippet.
+ * @param  {string} content Content in source format.
+ * @return {string}         Content in target format.
+ */
+function formatHtml(content) {
+    const ABSTRACT_PREFIX = '<section class="prog__container">';
+    const ABSTRACT_POSTFIX = '</section>';
+
+    return ABSTRACT_PREFIX + markdown.toHTML(content) + ABSTRACT_POSTFIX;
+}
```
Comment (pjhampton) [inline lib/fathead/eslint/lib/convert.js]: Hmm, this is was a good approach. Nice thinking!

**Comment by zodiac403** [comment]
@bfmags What is your oppinion? Did I miss anything?
Otherwise my PR is ready.

File: lib/fathead/eslint/trigger_words.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+eslint
```
Comment (moollaza) [inline lib/fathead/eslint/trigger_words.txt]: perhaps also `eslint rule` ?

Comment (zodiac403) [inline lib/fathead/eslint/trigger_words.txt]: *done*

**Comment by zodiac403** [comment]
Good points, @bfmags. I made article texts full sentences by starting with uppercase letters and ending with a dot. The titles I kept lowercase, as this is how they usually are referenced in the docs and in IDEs.

Some key words do trigger locally in the **duckpan query** and **duckpan server** environment &ndash; but don't trigger ont the beta server. I already posted this phenomena in the [forum](https://forum.duckduckhack.com/t/fathead-display-how-to-set-more-at-and-subtitle/1575/11?u=zodiac403). As **duckpan server** triggers I am not really sure, if I can do anything about them... Any idea?

At least I updated the IA page to contain only "green" examples.

**Comment by bfmags** [comment]
Thanks for the changes @zodiac403 !

@moollaza Any idea on the non triggering keywords?

e.g `eslint brace-style` https://beta.duckduckgo.com/?q=eslint%20brace-style
`brace-style	A			Stylistic Issues						http://eslint.org/img/logo.svg	<section class="prog__container"><p>Enforce consistent brace style for blocks.</p></section>	http://eslint.org/docs/rules/brace-style`

e.g. `stylistic issues : line-comment-position, etc` https://beta.duckduckgo.com/228/c/Stylistic_Issues?ia=list&iai=%2F228%2Flines-around-comment

**Comment by moollaza** [comment]
> Any idea on the non triggering keywords?

I'm not too sure why they're not working. Internally the data looks like it should trigger correctly on those queries.

Internally we replace dashes in the title with a space (` `) to build the "search_title", which is the primary lookup string. However, we maintain the original title, and do a fallback lookup when nothing matches on a `search_title`.

We also build a redirect with the dashes remove, so `brace-style` becomes `bracestyle` and that acts as another `search_title`:

https://beta.duckduckgo.com/?q=eslint+bracestyle&ia=about

For now this is complete though and ready for merge. We may have to make some internal changes to ensure it triggers correctly.

**Comment by zodiac403** [comment]
Hi there,
any new insights on the triggering issue?

- Searching only the rule title (e.g. **nounreachable**) does trigger the fathead: https://duckduckgo.com/?q=nounreachable&ia=about
- Working with the trigger word (**eslint nounreachable**) does not trigger: https://duckduckgo.com/?q=eslint+nounreachable&ia=web

Any config that I missed?

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=eslint no-unreachable)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=?5g6eixcMwH)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=eslint no-unreachable)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=?rJxjzg8oyN)

**Pull request merged**